first draft of unet training script

hajeressef opened the pull request:
PR to add training script of Unet architecture for instance segmentation. 
The abstract_trainer used can be found in [this PR](https://github.com/picselliahq/picsellia-training-engine/pull/17)

picsalex commented:
Thanks for the PR @hajeressef !
Some details I've noticed on the structure:
- The main folder `training-unet-instance-segmentation/...` is not consistent with the repo (`captured-...`)
- There's a remaining `pyproject.tom` file, I guess this one can be removed since we don't use poetry within the `base` image.
- There's a `.ipynb` file that can also be removed from the PR.

picsalex commented on training-unet-instance-segmentation/picsellia/utils.py:
```diff
@@ -0,0 +1,286 @@
+
+
+def get_training_augmentation() -> A.Compose:
+    train_transform = [
```
I think we need to think about a strategy for data augmentation in the future, I guess defining a hard-coded augmentation list might not suit the user's use-case. Maybe something like an `augmentation_config.json`, something the user can change when strating the container instead of rebuilding the image. idk what do you think @PN-picsell?

picsalex replied:
I guess we can mount a volume between the container and the host where the config files are stored. We can also provide a argument to specify with configuration file use in case there's several data_augmentation files. Something like:

```bash
docker run -v /training-unet-instance-segmentation/data_augmentation_files/:/data_augmentation_files picsellia/training-unet:latest
--data_augmentation_file_path /data_augmentation_files/config.json 
```

Maybe a strategy that looks like this:
- If there's no config file, or the config file path is wrong, use a default configuration
- Otherwise, load the config

We just need to think the best way to store those augmentations. In a `.py` file ? a `.json` file ?
I think this volume argument should stay optional and only be used when the user needs to experiment several data augmentations without rebuidling the docker image.

hajeressef commented:
Hi @picsalex . Thank you so much for your feedback on the structure. 

- for the naming I think we will rethink the renaming of all repos. I removed the captured just to seperate the new refactored directories from the old ones. Since there was also a yolov8 one that I refactored before so it needed a different name. 
- @PN-picsell asked me to add poetry files. do you mean that we would need to add something to the base image to take these into account?
- @PN-picsell and I discussed adding jupyter notebooks for future reference, in case we need to do future experiments. We could keep them here or think of another location (an outside folder) where we group them all.

picsalex commented:
- Okay, makes sense. I guess we could work on another PR to update the folder names.
- Okay, so can I guess we can just add them in the `.gitignore`?
- I think we should put them somewhere else because right now they are included in the Docker images during the build process. Maybe at the root in `training-unet-instance-segmentation/`?

hajeressef commented on training-unet-instance-segmentation/picsellia/trainer.py:
```diff
@@ -0,0 +1,226 @@
+from abstract_trainer.trainer import AbstractTrainer
```
Yes I have black enabled. It does not reorder the imports. I can reorder the imports as such: 
```
import os
import segmentation_models as sm
import keras
from picsellia.exceptions import ResourceNotFoundError

from abstract_trainer.trainer import AbstractTrainer
from utils import (
    split_train_test_val_filenames,
    makedirs_images_masks,
    move_images_and_masks_to_directories,
    Dataset,
    get_training_augmentation,
    get_preprocessing,
    get_validation_augmentation,
    Dataloader,
    format_and_log_eval_metrics,
    download_image_mask_assets,
    get_classes_from_mask_dataset,
)
```
pls feel free to suggest another way to make it look less messy

hajeressef commented on training-unet-instance-segmentation/picsellia/trainer.py:
```diff
@@ -0,0 +1,226 @@
+    get_classes_from_mask_dataset,
+)
+
+os.environ["SM_FRAMEWORK"] = "tf.keras"
```
Yes! I moved it to the dockerfile.

hajeressef commented:
1. We could perhaps keep the old ones for now with the old naming, and slowly clean up/delete the "captured" repos as they get reworked/refactored into new repos with new names.
2. no I believe we need it to be pushed. The idea is that whoever pulls this repo to experiment more or test without a docker image, can use this poetry file in their environment
3. Makes sense. I will move it to root directory as you mentioned.

picsalex commented:
- Okay, maybe can you update the README to mention this ? Or create an issue so we don't forget what we discussed about that matter.
- Same thing, maybe explain this in the README? I think we must be careful here since the dependencies are defined at 2 different places (requirements.txt and pyproject.toml).
- 👍

PN-picsell commented:
Looks really nice to me, @picsalex @hajeressef we can meet online to discuss the matters adressed here.
Just one think before we merge, can you update the notebook so it looks exactly like the main.py file ? This way it's clear where to start the research back
